### PR TITLE
Sets entities

### PR DESCRIPTION
**Pull request by WhiteBlackGoose**
As planned, we are going to implement `Set` as the part of the AM ecosystem: as `Entity`.

### Possible migration problems:
1. `Contains` => `ContainsNode`

### CODE REVIEW

File: AngouriMath/Core/Antlr/AngouriMath.g [hunk 1]
```diff
@@ -194,8 +196,22 @@ atom returns[Entity value]
     | 'arcsec(' args = function_arguments ')' { Assert("arcsec", 1, $args.list.Count); $value = MathS.Arcsec($args.list[0]); }
     | 'arccosec(' args = function_arguments ')' { Assert("arccosec", 1, $args.list.Count); $value = MathS.Arccosec($args.list[0]); }
     | 'gamma(' args = function_arguments ')' { Assert("gamma", 1, $args.list.Count); $value = MathS.Gamma($args.list[0]); }
-    | 'derivative(' args = function_arguments ')' { Assert("derivative", 3, $args.list.Count); $value = MathS.Derivative($args.list[0], $args.list[1], $args.list[2]); }
-    | 'integral(' args = function_arguments ')' { Assert("integral", 3, $args.list.Count); $value = MathS.Integral($args.list[0], $args.list[1], $args.list[2]); }
+    | 'derivative(' args = function_arguments ')' 
+        { 
+            Assert("derivative", 3, $args.list.Count);
```
Comment (Happypig375) [inline AngouriMath/Core/Antlr/AngouriMath.g:201]: Optional 3rd argument like `log`

Comment (WhiteBlackGoose) [reply AngouriMath/Core/Antlr/AngouriMath.g:201]: Hm, I guess why not

File: AngouriMath/Core/Domains.cs [hunk 2]
```diff
@@ -47,10 +47,10 @@ public static bool FitsDomainOrNonNumeric(Entity entity, Domain domain)
             (Domain.Any, "AA")
```
Comment (Happypig375) [inline AngouriMath/Core/Domains.cs:47]: Universal set (UU)?

File: AngouriMath/Core/FiniteSetBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿using AngouriMath.Core.Exceptions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using static AngouriMath.Entity;
+using static AngouriMath.Entity.Set;
+
+namespace AngouriMath.Core
+{
+    // TODO: is it useful for the user to have this class?
```
Comment (Happypig375) [inline AngouriMath/Core/FiniteSetBuilder.cs:10]: Just use a list??

File: Tests/UnitTests/Convenience/LatexTest.cs [hunk 4]
```diff
@@ -174,16 +175,16 @@ public class LatexTests
...
         [Fact] public void Set2() => Test(@"\left\{1,2\right\}", MathS.Sets.Finite(1, 2));
         [Fact] public void Set3() => Test(@"\left\{\sqrt{x},{x}^{2},\sin\left(x\right)\right\}", MathS.Sets.Finite(MathS.Sqrt(x), MathS.Sqr(x), MathS.Sin(x)));
-        [Fact] public void SetR() => Test(@"\left\{\left(-\infty ,\infty \right)\right\}", MathS.Sets.R());
+        [Fact] public void SetR() => Test(@"\left\{\left(-\infty ,\infty \right)\right\}", MathS.Sets.R);
         [Fact] public void SetC() =>
-            Test(@"\left\{\left\{z\in\mathbb C:\Re\left(z\right)\in\left(-\infty ,\infty \right)\wedge\Im\left(z\right)\in\left(-\infty ,\infty \right)\right\}\right\}", MathS.Sets.C());
+            Test(@"\left\{\left\{z\in\mathbb C:\Re\left(z\right)\in\left(-\infty ,\infty \right)\wedge\Im\left(z\right)\in\left(-\infty ,\infty \right)\right\}\right\}", MathS.Sets.C);
         [Fact] public void SetIntervalCloseClose() =>
-            Test(@"\left\{\left[1,2\right]\right\}", new Set(MathS.Sets.Interval(1, 2)));
+            Test(@"\left\{\left[1,2\right]\right\}", (1, 2).ToEntity());
```
Comment (Happypig375) [inline Tests/UnitTests/Convenience/LatexTest.cs:187]: Despite having parentheses, the corresponding interval is closed on both sides?

File: AngouriMath/Core/Domains.cs (hunk 2 above)
Comment (WhiteBlackGoose) [reply AngouriMath/Core/Domains.cs:47]: Universal set is specific some area. Say, in math we usually refer to U as to the whole set of reals, complexes (depends on the context). Here it means literally anything 🦆

File: AngouriMath/Core/FiniteSetBuilder.cs (hunk 3 above)
Comment (WhiteBlackGoose) [reply AngouriMath/Core/FiniteSetBuilder.cs:10]: Having a separate type might be better, if something needs to be added

File: Tests/UnitTests/Convenience/LatexTest.cs (hunk 4 above)
Comment (WhiteBlackGoose) [reply Tests/UnitTests/Convenience/LatexTest.cs:187]: Yes :).

File: AngouriMath/Core/Domains.cs (hunk 2 above)
Comment (Happypig375) [reply AngouriMath/Core/Domains.cs:47]: https://en.wikipedia.org/wiki/Universal_set
> In set theory, a universal set is a set which contains all objects, including itself.

It really just depends on what you can work with.

File: AngouriMath/Convenience/AngouriMathExtensions.cs [hunk 5]
```diff
@@ -41,15 +41,15 @@ public static FiniteSet ToSet(this IEnumerable<Entity> elements)
         /// </summary>
         /// <returns>A set of unique elements</returns>
         public static Set Unite(this IEnumerable<Set> sets)
-            => sets.Aggregate((a, b) => MathS.Union(a, b));
+            => sets.Any() ? sets.Aggregate((a, b) => MathS.Union(a, b)) : Empty;
```
Comment (Happypig375) [inline AngouriMath/Convenience/AngouriMathExtensions.cs:44]: Empty can be the seed: https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.aggregate?view=netcore-3.1#System_Linq_Enumerable_Aggregate__2_System_Collections_Generic_IEnumerable___0____1_System_Func___1___0___1__

File: AngouriMath/Functions/Continuous/Solvers/EquationSolver/AnalyticalEquationSolver.cs [hunk 6]
```diff
@@ -144,7 +144,7 @@ static Entity TryDowncast(Entity equation, Variable x, Entity root)
                         continue;
                     // Here we are trying to solve for this replacement
                     var solutionsSet = Solve(alt.Substitute(minimumSubtree, newVar), newVar);
-                    if (solutionsSet is FiniteSet enums)
+                    if (solutionsSet is FiniteSet enums && !enums.IsSetEmpty)
```
Comment (Happypig375) [inline AngouriMath/Functions/Continuous/Solvers/EquationSolver/AnalyticalEquationSolver.cs:147]:
```suggestion
                    if (solutionsSet is FiniteSet { IsSetEmpty:false } enums)
```

File: Tests/UnitTests/Algebra/SolveTest/SolveOneEquation.cs [hunk 7]
```diff
@@ -119,10 +117,9 @@ public void InvertedFunctions(string func, int rootAmount)
         {
             Entity toRepl = func + "(x2 + 3)";
             Entity expr = MathS.Sqr(toRepl) + 0.3 * toRepl - 0.1 * MathS.Var("a");
-            var roots = expr.SolveEquation(x).InnerSimplified;
-            Assert.IsType<FiniteSet>(roots);
-            if (roots is not FiniteSet finite)
-                throw new AngouriBugException("Eeem?");
+            var roots = expr.SolveEquation(x);
+            roots = (Set)roots.InnerSimplified;
+            var finite = Assert.IsType<FiniteSet>(roots);
```
[the remark this answers is not in the document]
Comment (WhiteBlackGoose) [reply Tests/UnitTests/Algebra/SolveTest/SolveOneEquation.cs:122]: It was for debugging

File: AngouriMath/Functions/Evaluation/Evaluation.Definition.cs [hunk 8]
```diff
@@ -28,6 +29,10 @@ namespace AngouriMath
         /// </summary>
         internal abstract Entity InnerSimplify();
 
+        /// <summary>
+        /// This is the result of naive simplifications. In other 
+        /// symbolic algebra systems it is called "Automatic simplification"
```
[the remark this answers is not in the document]
Comment (WhiteBlackGoose) [reply AngouriMath/Functions/Evaluation/Evaluation.Definition.cs:34]: Those are not numerical simplifications though

**Pull request merged**